mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,207 @@
+
+    @typed_pos_args(
+        "gobuild.project",
+        str,
+        varargs=(
+            str,
+            mesonlib.File,
+            build.CustomTarget,
+            build.CustomTargetIndex,
+            build.GeneratedList,
+        ),
+    )
+    @typed_kwargs(
+        "gobuild.project",
+        CT_INPUT_KW.evolve(name="sources"),
+        KwargInfo(
+            "include_directories",
+            ContainerTypeInfo(list, (str, build.IncludeDirs)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
+        KwargInfo(
+            "gomod_tidy",
+            bool,
+            default=True,
+        ),
+    )
+    def project(
+        self,
+        state: ModuleState,
+        args: T.Tuple[
+            str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        ],
+        kwargs: ProjectKwargs,
+    ) -> ModuleReturnValue:
+        if self.go_exe is None:
+            self.detect_tools(state, kwargs["native"])
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs["sources"]))
+        )
+        self.counter += 1
+
+        go_sum_target = build.CustomTarget(
+            f"go mod tidy {self.counter}",
+            state.subdir,
+            state.subproject,
+            state.environment,
+            [
+                self.go_exe,
+                "-C",
+                "@CURRENT_SOURCE_DIR@",
+                "mod",
+                "tidy" if kwargs["gomod_tidy"] else "verify",
+            ],
+            self.interpreter.source_strings_to_files(["go.mod"]),
+            [f"go-mod-tidy-{self.counter}.log"],
+            capture=True,
+            build_always_stale=True,
+        )
+
+        build_dir_abs = Path(state.environment.get_build_dir())
+        target_dir_abs = build_dir_abs / state.backend.get_target_dir(go_sum_target)
+
+        link_dirs = {
+            "-L" + (build_dir_abs / state.backend.get_target_dir(lib)).as_posix()
+            for lib in kwargs["link_with"]
+        }
+        link_targets = {"-l" + lib.name for lib in kwargs["link_with"]}
+
+        if kwargs["native"] or not state.environment.is_cross_build():
+            extra_env = {}
+        else:
+            extra_env = {
+                "GOOS": "windows",
+                "GOARCH": "amd64",
+                "CGO_ENABLED": "1",
+                "CC": "x86_64-w64-mingw32-gcc",
+                "CXX": "x86_64-w64-mingw32-g++",
```

Comment:
Help wanted to retrieve the compiler commands, e.g. `ccache {my_gcc_path}` from the meson cross-compilation config `cross-file.ini`.

Reply:
Also needed to generate a step `compiler.has_argument('-mthreads`)` because `go build` and CGO assumes gcc toolchain.

Reply:
> Here's another place for that native keyword

> `cc = state.environment.coredata.compilers[kwargs['native']]['c']`
> This gives you the Compiler object backing the DSL compiler, which gives you access to things like >has_argument and what the argument syntax for linking is. This also has the methods to get the arguments for the C compiler you were looking for later.

> While the Linux -> Windows cross makes a ton of sense, I'm hesitant to assume no one is ever going to do Windows ->Windows compilation.

Resolved in the new commit. Closing.